### PR TITLE
Adding Final energy per carrier tag

### PR DESCRIPTION
**Pull request by Renato-Rodrigues**
This pull request
-  adds the {Final Energy carrier} tag to support NAVIGATE template variables.
- change the energy-final.yaml file to:  
   * use the FE tag whenever possible
   * move variables that do not belong to the NAVIGATE template to the end of the file. 

While working on this, it became evident that there is one design decision that needs to be made regarding the use of the tag system.
As you can see in this pull request, I made use of the {FE carrier} tag as much as possible for the FE variables definition. This creates additional variables that will never have an assigned value, e.g. `Final Energy|Transport|{Final Energy carrier}` creates the variable `Final Energy|Transport|Heat` that will never have a value; the same for `Final Energy|Non-Energy Use|Electricity`, and so on. 

There are three options to handle that as far as I can see:
1. keep using the tag system as it is done in this pull request. Trade-off: This simplifies the variables list definition, reducing entries, description repetitions and possible variables missing mistakes. However, this also creates additional variables in the database that will never have any values assigned.
2. add support in the tag creation to define subgroups of the complete tag list. This would require additional code and an additional "sub-tag" field in the variable definition contained in the tag yaml file. With this tag we could filter out only part of the tag variables for these special cases, instead of using the full list of variables contained in the tag. This would still have all advantages of alternative 1, but it would allow to not create any unnecessary variable. Alternatively, we could also create as many additional tags as necessary to avoid meaningless variables, but this would add a lot of unnecessary repetition in descriptions and variable names.
3. revert to use the traditional and more extensive list of variables, limiting the tags use only for some handful of cases.

I would lean to the option 2, but as this is dependable of some code under the hood I would say this is entirely your decision @danielhuppmann
Please let me know how should I proceed to add the NAVIGATE template variables.

### CODE REVIEW

**Comment by Renato-Rodrigues** [comment]
Additional tags added to support the full list of Final Energy variables in the NAVIGATE template:
* {Transport mode} - list of transport modes and transport mode energy carriers
* {Industry sectors} - navigate industry sectors 
* {Bunkers mode} - full list of bunker modes and bunker mode energy carriers

**Review by phackstock — COMMENTED** [review]
Thanks @Renato-Rodrigues for pushing this forward.
Looks good to me overall, please find one comment below and one more general remark.
Regarding the question you raised about unintended additional variables. This is a tricky one and perhaps best to be converted into an issue on https://github.com/iamconsortium/nomenclature. 
My first quick take hinges on a follow-up question. Are these unintended additional variables nonsensical, i.e. would no model ever report them or could they exist but are to be excluded from this model comparison exercise.
If its the former, I would prefer the first option as the codelist just defines what is allowed and does not create any entries the data base or a scenario explorer documentation. If it is the latter I'd agree with you that we should implement some kind of exclusion list. I'm wondering on what side it would be better, the tag or the variable that uses the tag ...
The reason why I'd go with an exclusion list rather than an inclusion list is based on the assumptions that if we have a tag we would probably want to use the majority and exclude a smaller subset thereby reducing the amount of explicit definitions required. 
In short, I think that's a great discussion to have over at https://github.com/iamconsortium/nomenclature in a dedicated issue.

File: definitions/variable/tag_fe_carriers.yaml [hunk 1]
```diff
@@ -0,0 +1,95 @@
+# This list of inputs will replace any occurrence of `{Final Energy carrier}`
+# in a variable-definition yaml file
+
+- Final Energy carrier:
+  - Electricity:
+      description: electricity (including on-site solar PV), excluding transmission/distribution losses
+      weight: Electricity
```
[the remark this answers is not in the document]
Comment (Renato-Rodrigues) [reply definitions/variable/tag_fe_carriers.yaml:7]: The intention was to allow the weight definition for price variables: https://github.com/openENTRANCE/openentrance/pull/166#issuecomment-1080769555

For example:
```
- Price|Secondary Energy|{Final Energy carrier}:
    description: price at the secondary level of {Final Energy carrier}, i.e. for large scale
                consumers (e.g. aluminum production, gas power plant)
    notes: Prices should include the effect of carbon prices.
    unit: [EUR_2020/GJ, USD_2010/GJ]
    weight: Secondary Energy|{Final Energy carrier}
```
All quantity variables contained in this pull request does not require any weight by definition. This only added to support the weight overwrite on cost and price cariables from other pull requests.

[the remark this answers is not in the document]
Comment (phackstock) [reply definitions/variable/tag_fe_carriers.yaml:7]: Ah ok, then all good, never mind.

[the remark this answers is not in the document]
Comment (Renato-Rodrigues) [reply definitions/variable/tag_fe_carriers.yaml:7]: Regarding the unintended additional variables, I definitely agree with you about the exclusion list being a more viable option. 

I would say that we have a hybrid of nonsensical variables and variables excluded due to tractability reasons.

Below some examples of nonsensical variables:

- ex 1. all electricity use is energy related, so it will never make sense to have the variable `Final Energy|Non-Energy Use|Electricity` as far as I know. 
- ex 2. The {Final Energy carrier} tag was not used to create the emission variables in the emissions pull request (https://github.com/openENTRANCE/openentrance/pull/163) because it would create a lot of nonsensical emission variables, such as `Emissions|CO2|Energy|Demand|Transportation|Electricity`, which is clearly wrong as BEVs don't emit CO2 when using the electricity in the bateries. 

However, there are some variables that are semantically perfect fine, but are not included in the current variables templates like NAVIGATE to avoid unnecessary duplications (see example 3 below), or just for practicality of the cariables template size (example 4):

- ex 3. `Transportation|2W and 3W|Passenger` is just a duplication of `Transportation|2W and 3W`, so to avoid exploding the number of variables reported, models and variables templates only include `Transportation|2W and 3W` usually.
- ex 4. `Emissions|CO2|Energy|Demand|Transportation|{Transport mode, tech and carrier}`

File: definitions/variable/energy/energy-final.yaml [hunk 2]
```diff
@@ -19,409 +19,361 @@
...
-- Final Energy|Fossil:
-    description: Final energy consumption from fossil fuels
-    energy source: fossil fuels
+- Final Energy (w/o bunkers):
+    description: Total final energy consumption by all end-use sectors and all fuels,
+      excluding transmission/distribution losses, excluding bunker fuels, excluding pipelines 
+    energy source: all
     energy use: all excl. losses
     sector: all end-users
     unit: [EJ/yr, GWh]
```
[the remark this answers is not in the document]
Comment (Renato-Rodrigues) [reply definitions/variable/energy/energy-final.yaml:28]: I think we had this discussion before, maybe it didn't included you at the time. 
If my memory does not fail me, it was decided that multiple units are allowed, however at each project all result submissions should only upload results using a single unit standard to avoid any issues.

[the remark this answers is not in the document]
Comment (phackstock) [reply definitions/variable/energy/energy-final.yaml:28]: Ah right I see, I assume that this has to do with the definitions being used in openENTRANCE as well as ECEMF?
This approach is fine, however it is a little dangerous as there is no safeguard against mixed unit uploads. 
Maybe we want to implement a project specific check in `workflow.py`.

File: definitions/variable/tag_fe_carriers.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (phackstock) [reply definitions/variable/tag_fe_carriers.yaml:7]: Ok right, thanks for the quick clarification, then this definitely sounds like an issue/feature request for the nomenclature package.
The main question would be on which side we should implement this, the tag side or the variable side, but we can move this discussion over to a dedicated issue.
